### PR TITLE
Passive DCC

### LINKED ISSUE #487: Passive DCC
Currently weechat only allows "active" DCC.
I have found a request on the old savannah nongnu project page, but nowhere else.
So I create a new one here on github.

http://savannah.nongnu.org/task/?5572

### PR DESCRIPTION
**Pull request by mario-campos**
This PR adds initial support for [passive DCC](https://en.wikipedia.org/wiki/Direct_Client-to-Client#Reverse_/_Firewall_DCC). Specifically, it supports a non-WeeChat client sending `DCC SEND` with a port of 0, which causes weechat to open a listening socket, accept the first client that connects, and start a new xfer.

- It does (or should) respect the `xfer.network.own_ip` config property.
- It does (or should) respect the `xfer.network.port_range` config property.
- It does (or should) support the passive DCC RESUME extension (linked above).

Closes #487.

### CODE REVIEW

File: src/plugins/irc/irc-command.c [hunk 1]
```diff
@@ -2148,7 +2148,7 @@ IRC_COMMAND_CALLBACK(dcc)
             {
                 weechat_infolist_new_var_string (item, "plugin_name", weechat_plugin->name);
                 weechat_infolist_new_var_string (item, "plugin_id", ptr_server->name);
-                weechat_infolist_new_var_string (item, "type_string", "file_send");
+                weechat_infolist_new_var_string (item, "type_string", "file_send_passive");
```
Comment (flashcode) [inline src/plugins/irc/irc-command.c:2151]: "file_send_passive" is the same behavior as before here?
Quite confusing for me to read "passive" as it's not, or am I wrong?
So it seems there's no way for WeeChat to initiate both DCC types? I mean for example `/dcc send` (legavy behavior) and `/dcc sendpassive` (new)?

File: src/plugins/irc/irc-ctcp.c [hunk 2]
```diff
@@ -862,7 +932,9 @@ irc_ctcp_recv_dcc (struct t_irc_server *server, const char *nick,
                 weechat_infolist_new_var_string (item, "proxy",
                                                  IRC_SERVER_OPTION_STRING(server, IRC_SERVER_OPTION_PROXY));
                 weechat_infolist_new_var_string (item, "remote_address", pos_addr);
+                weechat_infolist_new_var_string (item, "local_address", str_address);
                 weechat_infolist_new_var_integer (item, "port", atoi (pos_port));
+                weechat_infolist_new_var_string (item, "token", pos_token);
```
Comment (flashcode) [inline src/plugins/irc/irc-ctcp.c:937]: `token` field should be an integer here:

```C
                weechat_infolist_new_var_integer (item, "token", atoi (pos_token));
```

File: src/plugins/irc/irc-ctcp.c [hunk 3]
```diff
@@ -1050,11 +1215,12 @@ irc_ctcp_recv_dcc (struct t_irc_server *server, const char *nick,
             {
                 weechat_infolist_new_var_string (item, "plugin_name", weechat_plugin->name);
                 weechat_infolist_new_var_string (item, "plugin_id", server->name);
-                weechat_infolist_new_var_string (item, "type_string", "file_recv");
+                weechat_infolist_new_var_string (item, "type_string", "file_recv_active");
                 weechat_infolist_new_var_string (item, "filename",
                                                  (filename) ? filename : pos_file);
                 weechat_infolist_new_var_integer (item, "port", atoi (pos_port));
                 weechat_infolist_new_var_string (item, "start_resume", pos_start_resume);
+                weechat_infolist_new_var_string (item, "token", pos_token);
```
Comment (flashcode) [inline src/plugins/irc/irc-ctcp.c:1223]: `token` field should be an integer here:

```C
                weechat_infolist_new_var_integer (item, "token", atoi (pos_token));
```

File: src/plugins/xfer/xfer.c [hunk 4]
```diff
@@ -1039,6 +1061,7 @@ xfer_add_cb (const void *pointer, void *data,
     charset_modifier = weechat_infolist_string (infolist, "charset_modifier");
     filename = weechat_infolist_string (infolist, "filename");
     proxy = weechat_infolist_string (infolist, "proxy");
+    token = atoi (weechat_infolist_string (infolist, "token"));
```
Comment (flashcode) [inline src/plugins/xfer/xfer.c:1064]: `token` field should be an integer here:

```C
    token = weechat_infolist_integer (infolist, "token");
```

**Review by flashcode — CHANGES_REQUESTED** [review]
The `token` is never set by WeeChat itself so if I understand WeeChat only supports passive DCC as a "receiver".

The help on command `/dcc` (ie `/help dcc`) should maybe mention passive DCC being supported.

Besides, the unit tests are failing, because some network address resolution seems to be made and causes a test to fail.

File: src/plugins/xfer/xfer.c (hunk 4 above)
Comment (mario-campos) [reply src/plugins/xfer/xfer.c]: Done.

File: src/plugins/irc/irc-ctcp.c (hunk 3 above)
Comment (mario-campos) [reply src/plugins/irc/irc-ctcp.c]: Done.

File: src/plugins/irc/irc-ctcp.c (hunk 2 above)
Comment (mario-campos) [reply src/plugins/irc/irc-ctcp.c]: Done.

**Comment by mario-campos** [comment]
>The token is never set by WeeChat itself so if I understand WeeChat only supports passive DCC as a "receiver".

Correct.

>The help on command /dcc (ie /help dcc) should maybe mention passive DCC being supported.
>Besides, the unit tests are failing, because some network address resolution seems to be made and causes a test to fail.

I'll address these.

**Comment by mario-campos** [comment]
@flashcode, can you please point me to the file that contains the help docs (`/help dcc`)?

**Comment by mario-campos** [comment]
@flashcode, can you try running the test suite again? It should pass now.

**Comment by mario-campos** [comment]
@flashcode, have you been able to test this PR?

**Comment by flashcode** [comment]
I'll test it soon, in the next days.

**Review by flashcode — APPROVED** [review]

**Comment by flashcode** [comment]
Merged, thanks for adding this new feature!

**Comment by flashcode** [comment]
Hi @mario-campos,

For your information, this PR introduced a regression with DCC chat, causing a crash when you do `/dcc chat <nick>`.

I fixed the issue in this commit e62ff286ff4141e276f22b834b09aa72ac509799 and you could look if my changes are OK for you.
At least I tested the DCC chat which is now working again, but didn't fully retest the passive DCC.

**Comment by mario-campos** [comment]
Sorry about introducing that regression. I retested the passive DCC functionality against the HEAD of master (f49810ec0e3f20a8a9b44285b2ae59ae21db64ac) and it still works. Thanks!

**Comment by ksurl** [comment]
how should the port range be written in xfer.conf?

will a dash work? or does it use a colon

`port_range = "2000-2010"`

**Comment by flashcode** [comment]
@ksurl: it's a dash, see `/help xfer.network.port_range`:

```
description: restricts outgoing files/chats and incoming/passive files to use only ports
in the given range (useful for NAT) (syntax: a single port, ie. 5000 or a port range,
ie. 5000-5015, empty value means any port, it's recommended to use ports greater than 1024,
because only root can use ports below 1024)
```

**Pull request closed without merging**